[NAV] Calibration stage

brano2 opened the pull request:
* Calibration phase calculations
* Barrier for syncing with motors
* Updating `NavigationState`s
* Reading state machine states

brano2 committed:
Create nav-mot synchronisation barrier

brano2 committed:
Implement calibration period & internal states; sync with motors

brano2 committed:
Check and react to stm state changes

brano2 committed:
Fix lint errors

brano2 committed:
Move nav-mot barrier to system.hpp

brano2 committed:
Refactor navigation

Move the nav-mot barrier to constructor from the finishCalibration
function

brano2 committed:
Merge branch 'develop' into nav-calibration

brano2 committed:
Add transition from calibrating to ready

brano2 committed:
Change stm transition detection

It seems like a better idea to check just the current state as opposed
to transition. If calibration is not successfully concluded on stm
transition to 'accelerating' for whatever reason, it will be attempted
again this way.

mkristien commented on BeagleBone_black/src/navigation/main.cpp:
```diff
@@ -41,6 +44,14 @@ void Main::run()
+    if (current_state != State::kIdle
+        && current_state != State::kReady
+        && nav_.getState() != NavigationState::kOperational
+        && nav_.getState() != NavigationState::kCriticalFailure)
```
I would advice specifying in which nav states the finishCalibration is possible, rather than in which it is not. e.g. have only `nav_.getState() == NavigationState::kReady` as the condition.

mkristien commented on BeagleBone_black/src/navigation/main.cpp:
```diff
@@ -41,6 +44,14 @@ void Main::run()
+        && current_state != State::kReady
+        && nav_.getState() != NavigationState::kOperational
+        && nav_.getState() != NavigationState::kCriticalFailure)
+      nav_.finishCalibration();
```
the return value is not used, how do you know if you called the barrier?

Can you separate the calibration loop (and this exit check) from the main operational loop. Those can be two separate whiles, so that it is clear the code cannot start calibrating again after already being operational.

brano2 replied:
It turns out I don't really care. If I didn't call the barrier, then I'm still calibrating. I thought it might be useful when I wrote the function, but it turned out, `finishCalibration()` could have been declared to return `void` as well.

mkristien commented on BeagleBone_black/src/navigation/navigation.cpp:
```diff
@@ -94,6 +133,20 @@ void Navigation::update(ImuArray imus, ProximityArray proxis, DataPoint<uint32_t
+void Navigation::calibrationUpdate(ImuArray imus)
+{
+  for (unsigned int i = 0; i < data::Sensors::kNumImus; ++i) {
+    g_ += imus[i].acc.value;
```
depending on the number of samples in the calibration phase, have you considered an online mean update strategy to have better precision? For integer types, I would be worried about overflow, but since you are using float, you might only lose resolution with high values.

brano2 committed:
Merge branch 'develop' into nav-calibration

brano2 committed:
Refactor condition for finishing calibration

brano2 committed:
Change mean calculation algorithm

mkristien commented on BeagleBone_black/src/navigation/navigation.cpp:
```diff
@@ -135,12 +135,13 @@ void Navigation::update(ImuArray imus, ProximityArray proxis, DataPoint<uint32_t
   for (unsigned int i = 0; i < data::Sensors::kNumImus; ++i) {
-    g_ += imus[i].acc.value;
-    gyro_offsets_[i] += imus[i].gyr.value;
+    ++num_gravity_samples_;
```
Why is `num_gravity_samples_` incremented in the loop body but `num_gyro_samples_` out of the loop?

mkristien committed:
Merge branch 'develop' into nav-calibration

mkristien commented:
sorry, the auto merge with develop seemed to have broken things

brano2 committed:
Adjust to refined stm

 - Add proper initialisation during Idle state
 - Start calibration in Calibrating state
 - Indicate `kReady` once enough calibration
 - Stop calibration in Accelerating state
 - Refactor updating Data in main (now in a function)
 - No checking for crit. failure yet, that will be a different PR

brano2 committed:
Merge branch 'develop' into nav-calibration

brano2 commented:
@mkristien can you review again please?

mkristien commented on BeagleBone_black/src/navigation/main.hpp:
```diff
@@ -28,6 +28,7 @@
 namespace hyped {
 
 using data::Sensors;
+using data::State;
```
if you are not using this in the header, consider moving the import into .cpp file

brano2 committed:
Move unused import

brano2 committed:
Merge branch 'develop' into nav-calibration
